### PR TITLE
Time entry log grouping logic

### LINKED ISSUE #21: Add logic for grouping time entries
## Description
We need to add the logic that groups similar time entries, similar to what [we have in the current app](https://github.com/toggl/mobileapp/blob/83418effb77e5957a68679c0ab2c4fe49ec36512/Toggl.Core.Tests/UI/Transformations/TimeEntriesFlatenningTests.cs). Since there's no switch to enable/disable grouping, we should leave grouping on by default until we start implementing the settings view

## Definition of done
- [ ] Similar time entries are turned into a group
- [ ] Tests ensure the grouping logic is correct

### PR DESCRIPTION
**Pull request by Odrakir**
I'm splitting #39 in two. This one takes care of the logic only. 

Closes #21 

It includes the tests as we had them in the previous app (minus the date formatting stuff which will be tested elsewhere when included)

The `toggleTimeEntryGroupTapped` action takes a groupId. Ids should be type safe, but I'm leaving that for another PR as it's a big change.

### CODE REVIEW

File: Modules/Timer/Timer/TimeEntriesLog/ViewModels/TimeEntryViewModel.swift [hunk 1]
```diff
@@ -24,34 +25,32 @@ public struct TimeEntryViewModel: Equatable {
...
+        self.groupId = timeEntry.groupId
         self.description = timeEntry.description.isEmpty ? "No description" : timeEntry.description
         self.projectTaskClient = getProjectTaskClient(from: project, task: task, client: client)
         self.billable = timeEntry.billable
                 
         self.start = timeEntry.start
         self.duration = timeEntry.duration >= 0 ? timeEntry.duration : nil
         if let duration = duration {
-            self.durationString = duration.formatted
+            self.durationString = duration.formattedDuration(ending: Date())
```
Comment (heytherewill) [inline Modules/Timer/Timer/TimeEntriesLog/ViewModels/TimeEntryViewModel.swift]: Shouldn't we be avoiding using raw dates like that?

File: Modules/Timer/Timer/TimeEntriesLog/ViewModels/DayViewModel.swift [hunk 2]
```diff
@@ -1,15 +1,30 @@
...
+        day = timeLogCells.first!.start.ignoreTimeComponents()
         dayString = day.toDayString()
-        self.timeEntries = timeEntries
+        self.timeLogCells = timeLogCells
+        
+        durationString = self.timeLogCells
+                .filter({ !$0.isInGroup })
+                .map({ $0.duration ?? 0 })
+                .reduce(0, +)
+                .formattedDuration(ending: Date())
```
Comment (heytherewill) [inline Modules/Timer/Timer/TimeEntriesLog/ViewModels/DayViewModel.swift]: Same comment about dates

File: Modules/Timer/Timer/TimeEntriesLog/TimeEntriesSelector.swift [hunk 3]
```diff
@@ -1,10 +1,69 @@
...
         .map(DayViewModel.init)
         .sorted(by: { $0.day > $1.day })
 }
+
+let groupTimeEntries: (Set<Int>) -> ([TimeEntryViewModel]) -> [TimeLogCellViewModel] = { expandedGroups in
+
+    return { timeEntryViewModels in
+        return timeEntryViewModels
+            .grouped { $0.groupId }
+            .sorted(by: { $0.first!.start > $1.first!.start })
```
Comment (heytherewill) [inline Modules/Timer/Timer/TimeEntriesLog/TimeEntriesSelector.swift]: Why not sort before grouping for a better looking syntax?

File: Modules/Utils/Utils/Extensions/TimeInterval+Extensions.swift [hunk 4]
```diff
@@ -0,0 +1,16 @@
+import Foundation
+
+extension TimeInterval {
+    public func formattedDuration(ending endingDate: Date) -> String {
```
Comment (heytherewill) [inline Modules/Utils/Utils/Extensions/TimeInterval+Extensions.swift]: This function should also be unit tested

**Comment by Odrakir** [comment]
I'll create an issue to deal with date formatting, because it's not part of this PR. About functional testing, this is still missing the UI PR. Coming up next.

File: Modules/Timer/Timer/TimeEntriesLog/TimeEntriesSelector.swift (hunk 3 above)
Comment (Odrakir) [reply Modules/Timer/Timer/TimeEntriesLog/TimeEntriesSelector.swift]: The intention was to sort the groups, not the individual entries. But anyway, this made me realise we were not sorting properly and we were not testing for the sorting either. I fixed that.

File: Modules/Timer/Timer/TimeEntriesLog/ViewModels/TimeEntryViewModel.swift (hunk 1 above)
Comment (Odrakir) [reply Modules/Timer/Timer/TimeEntriesLog/ViewModels/TimeEntryViewModel.swift]: As I said, let's do dates in another PR

File: Modules/Timer/Timer/TimeEntriesLog/ViewModels/DayViewModel.swift (hunk 2 above)
Comment (Odrakir) [reply Modules/Timer/Timer/TimeEntriesLog/ViewModels/DayViewModel.swift]: As I said, let's do dates in another PR

File: Modules/Utils/Utils/Extensions/TimeInterval+Extensions.swift (hunk 4 above)
Comment (Odrakir) [reply Modules/Utils/Utils/Extensions/TimeInterval+Extensions.swift]: As I said, let's do dates in another PR

File: Modules/Timer/TimerTests/TestUtils/TimeEntry+Extensions.swift [hunk 5]
```diff
@@ -0,0 +1,15 @@
+import Foundation
+import Models
+
+extension TimeEntry {
+    static func with(id: Int? = nil, description: String = "", start: Date? = nil, duration: TimeInterval? = nil, workspaceId: Int = 0) -> TimeEntry {
+        let timeEntryId = id ?? UUID().hashValue
```
Comment (dshr) [inline Modules/Timer/TimerTests/TestUtils/TimeEntry+Extensions.swift]: Wait, I thought we weren't doing UUIDs yet?

Comment (Odrakir) [reply Modules/Timer/TimerTests/TestUtils/TimeEntry+Extensions.swift]: It's really not important. I just need a random Int for the tests.

Comment (dshr) [reply Modules/Timer/TimerTests/TestUtils/TimeEntry+Extensions.swift]: Cool-cool

**Pull request merged**